cmd/spiffe-aws-assume-role/cli/cli.go

```diff
@@ -10,23 +11,34 @@ import (
 )
 
 type CredentialsCmd struct {
-	Audience       string `required:"" help:"SVID JWT Audience. Must match AWS configuration"`
-	SpiffeID       string `required:"" help:"The SPIFFE ID of this workload"`
-	WorkloadSocket string `optional:"" help:"Path to SPIFFE Workload Socket"`
-	RoleARN        string `required:"" help:"AWS Role ARN to assume"`
-	SessionName    string `optional:"" help:"AWS Session Name"`
-	STSEndpoint    string `optional:"" help:"AWS STS Endpoint variable"`
+	Audience        string        `required help:"SVID JWT Audience. Must match AWS configuration"`
+	SpiffeID        string        `required help:"The SPIFFE ID of this workload"`
+	WorkloadSocket  string        `optional help:"Path to SPIFFE Workload Socket"`
+	RoleARN         string        `required help:"AWS Role ARN to assume"`
+	SessionName     string        `optional help:"AWS Session Name"`
+	STSEndpoint     string        `optional help:"AWS STS Endpoint variable"`
+	SessionDuration time.Duration `optional type:iso8601duration help:"AWS session duration in ISO8601 duration format (e.g. PT5M for five minutes)"`
 }
 
-func (c *CredentialsCmd) Run() error {
+type CliContext struct {
+	JWTSourceProvider credentials.JWTSourceProvider
+	STSProvider       credentials.STSProvider
+}
+
+func (c *CredentialsCmd) Run(context *CliContext) error {
 	spiffeID, err := spiffeid.FromString(c.SpiffeID)
 	if err != nil {
 		return err
 	}
 
-	src := credentials.NewJWTSVIDSource(spiffeID, c.WorkloadSocket, c.Audience)
+	src := context.JWTSourceProvider(spiffeID, c.WorkloadSocket, c.Audience)
```

Comment:
In order to allow for mocking we've introduced some indirection: the JWTSourceProvider is a factory function (made available via our custom context) which masks the underlying creation logic.

Reply:
I don't think the factory pattern is very commonly used in Golang. This is fine but I think normally the main code would have a function defining its own JWTSourceProvider and the testing code would have a a function defining a test version.

Reply:
I have those functions but they're defined in the same file as JWTSourceProvider to promote reuse/discovery (the tests, at least, need to use the static version to supply a mock in multiple places).